File: src/core/resolver.js
import { cloneBoard, countFrozen } from './board.js';
import { isPowerup } from './cells.js';
import { findMatches, pickMergePositions, applyMerges } from './match.js';
import { applyGravityAndRefill } from './gravity.js';
import { applyPowerupEffect } from './powerup.js';

export function swapCells(board, from, to) {
  const tmp = board[from.r][from.c];
  board[from.r][from.c] = board[to.r][to.c];
  board[to.r][to.c] = tmp;
}

export function resolveBoard(board, swapFrom = null, swapTo = null, options = {}, layout = null, targetShapes = null) {
  const captureBoards = Boolean(options.captureBoards);
  let totalScore = 0;
  let chainScore = 0;
  let isFirstMerge = true;
  const specialGained = {};
  const steps = [];
  let hadMatch = false;

  while (true) {
    const matches = findMatches(board);
    if (matches.length === 0) break;
    hadMatch = true;

    const mergePositions = pickMergePositions(
      matches,
      isFirstMerge ? swapFrom : null,
      isFirstMerge ? swapTo : null
    );
    const merged = applyMerges(board, matches, mergePositions);
    totalScore += merged.score;
    if (!isFirstMerge) chainScore += merged.score;

    Object.entries(merged.specialGained).forEach(([shape, n]) => {
      specialGained[shape] = (specialGained[shape] || 0) + n;
    });

    steps.push({
      type: 'merge',
      score: merged.score,
      mergeEvents: merged.mergeEvents,
      specialGained: { ...merged.specialGained },
      unfrozen: merged.unfrozenByAdjacency,
      boardAfterMerge: captureBoards ? cloneBoard(board) : null,
    });
    applyGravityAndRefill(board, layout);
    steps.push({
      type: 'refill',
      boardAfterRefill: captureBoards ? cloneBoard(board) : null,
    });
    isFirstMerge = false;
  }

  return { totalScore, chainScore, specialGained, steps, hadMatch };
}

function resolvePowerupSwap(board, from, to, captureBoards = false, layout = null, targetShapes = null) {
  const first = board[from.r][from.c];
  const second = board[to.r][to.c];
  const powerups = [];
  if (isPowerup(first)) powerups.push({ pos: from, partner: to, cell: first });
  if (isPowerup(second)) powerups.push({ pos: to, partner: from, cell: second });

  if (powerups.length === 0) {
    return { usedPowerup: false, totalScore: 0, chainScore: 0, specialGained: {}, taskFromPowerup: {}, steps: [], hadMatch: false };
  }

  const combinedSpecial = {};
  const powerEvents = [];
  let totalScore = 0;
  let anyTriggered = false;

  for (const pw of powerups) {
    const { targets, specialGained, clearedByShape, upgraded, cleared } =
      applyPowerupEffect(board, pw.pos, pw.partner, layout, targetShapes);

    // 行/列/同道具：targets 可能为空（空行/列），但只要是道具就算触发
    anyTriggered = true;

    Object.entries(specialGained).forEach(([shape, n]) => {
      combinedSpecial[shape] = (combinedSpecial[shape] || 0) + n;
    });
    // bomb 或升级消除的格子才计分
    totalScore += cleared.length + Object.values(clearedByShape).reduce((a, b) => a + b, 0);

    powerEvents.push({
      powerupType: pw.cell.powerupType,
      shape: pw.cell.shape,
      targets,
      upgraded,
      cleared,
      triggerAt: pw.pos,
    });
  }

  if (!anyTriggered) {
    return { usedPowerup: false, totalScore: 0, chainScore: 0, specialGained: {}, taskFromPowerup: {}, steps: [], hadMatch: false };
  }

  const steps = [];
  steps.push({
    type: 'powerup',
    events: powerEvents,
    score: totalScore,
    boardAfterPowerup: captureBoards ? cloneBoard(board) : null,
  });

  applyGravityAndRefill(board, layout);
  steps.push({
    type: 'refill',
    boardAfterRefill: captureBoards ? cloneBoard(board) : null,
  });

  const chain = resolveBoard(board, from, to, { captureBoards }, layout, targetShapes);
  totalScore += chain.totalScore;

  Object.entries(chain.specialGained).forEach(([shape, n]) => {
    combinedSpecial[shape] = (combinedSpecial[shape] || 0) + n;
  });

  steps.push(...chain.steps);

  return {
    usedPowerup: true,
    totalScore,
    chainScore: chain.totalScore,
    specialGained: combinedSpecial,
    // 任务分统一只走 specialGained，taskFromPowerup 置空避免重复计分
    taskFromPowerup: {},
    steps,
    hadMatch: chain.hadMatch || true,
  };
}

export function trySwap(board, from, to, options = {}, layout = null, targetShapes = null) {
  const beforeFrozen = countFrozen(board);
  swapCells(board, from, to);
  const afterSwapBoard = options.captureBoards ? cloneBoard(board) : null;

  const powerRes = resolvePowerupSwap(board, from, to, options.captureBoards, layout, targetShapes);
  if (powerRes.usedPowerup) {
    return { success: true, afterSwapBoard, unfrozenCount: beforeFrozen - countFrozen(board), isPop: false, ...powerRes };
  }

  const normalRes = resolveBoard(board, from, to, options, layout, targetShapes);
  return {
    success: true,
    afterSwapBoard,
    taskFromPowerup: {},
    unfrozenCount: beforeFrozen - countFrozen(board),
    isPop: false,
    ...normalRes,
  };
}

/**
 * 捏爆：直接清空一个普通非冰冻格，不解冻相邻冰壳。
 * 清空后重力补位并结算连锁（连锁消除按正常规则解冻、计任务分）。
 */
export function popCell(board, r, c, options = {}, layout = null, targetShapes = null) {
  const cell = board[r][c];
  if (!cell || cell.kind !== 'normal' || cell.frozen) {
    return {
      success: false, usedPowerup: false, totalScore: 0, chainScore: 0,
      specialGained: {}, taskFromPowerup: {}, steps: [], hadMatch: false,
      unfrozenCount: 0, isPop: true, popped: false,
    };
  }

  const beforeFrozen = countFrozen(board);
  board[r][c] = null; // 清空，不触发相邻解冻
  const steps = [];
  if (options.captureBoards) {
    steps.push({ type: 'pop', popped: { r, c }, boardAfterPop: cloneBoard(board) });
  }
  applyGravityAndRefill(board, layout);
  if (options.captureBoards) {
    steps.push({ type: 'refill', boardAfterRefill: cloneBoard(board) });
  }

  const chain = resolveBoard(board, null, null, options, layout, targetShapes);
  steps.push(...chain.steps);

  return {
    success: true,
    usedPowerup: false,
    totalScore: chain.totalScore,
    chainScore: chain.chainScore,
    specialGained: chain.specialGained,
    taskFromPowerup: {},
    steps,
    hadMatch: chain.hadMatch,
    // 捏爆本身不解冻；解冻仅来自连锁消除
    unfrozenCount: beforeFrozen - countFrozen(board),
    isPop: true,
    popped: true,
  };
}

export function simulateSwap(board, from, to, options = {}, layout = null, targetShapes = null) {
  const sim = cloneBoard(board);
  const result = trySwap(sim, from, to, options, layout, targetShapes);
  if (options.includeFinalBoard) return { ...result, finalBoard: cloneBoard(sim) };
  return result;
}

/** 捏爆的模拟版本（前端预演用）。 */
export function simulatePop(board, r, c, options = {}, layout = null, targetShapes = null) {
  const sim = cloneBoard(board);
  const result = popCell(sim, r, c, options, layout, targetShapes);
  if (options.includeFinalBoard) return { ...result, finalBoard: cloneBoard(sim) };
  return result;
}

File: rl_python/match3_engine/resolver.py
from __future__ import annotations

import random
from typing import Optional

from .board import Board, clone_board
from .cells import is_powerup
from .match import find_matches, pick_merge_positions, apply_merges
from .gravity import apply_gravity_and_refill
from .powerup import powerup_upgrade_targets


def swap_cells(board: Board, fr: dict, to: dict) -> None:
    board[fr["r"]][fr["c"]], board[to["r"]][to["c"]] = board[to["r"]][to["c"]], board[fr["r"]][fr["c"]]


def resolve_board(
    board: Board,
    rng: random.Random,
    swap_from: Optional[dict] = None,
    swap_to: Optional[dict] = None,
    layout: Optional[list] = None,
    target_shapes: Optional[list] = None,
) -> dict:
    total_score = 0
    chain_score = 0
    is_first = True
    special_gained: dict = {}
    cleared_by_shape: dict = {}
    merge_events: list = []
    had_match = False

    while True:
        matches = find_matches(board)
        if not matches:
            break
        had_match = True
        merge_positions = pick_merge_positions(
            matches,
            swap_from if is_first else None,
            swap_to if is_first else None,
        )
        merged = apply_merges(board, matches, merge_positions, rng)
        total_score += merged["score"]
        if not is_first:
            chain_score += merged["score"]
        for shape, n in merged["special_gained"].items():
            special_gained[shape] = special_gained.get(shape, 0) + n
        for shape, n in merged["cleared_by_shape"].items():
            cleared_by_shape[shape] = cleared_by_shape.get(shape, 0) + n
        merge_events.extend(merged.get("merge_events", []))
        apply_gravity_and_refill(board, rng, layout)
        is_first = False

    return {
        "total_score": total_score,
        "chain_score": chain_score,
        "special_gained": special_gained,
        "cleared_by_shape": cleared_by_shape,
        "merge_events": merge_events,
        "had_match": had_match,
        "used_powerup": False,
        "task_from_powerup": {},
    }


def _resolve_powerup_swap(
    board: Board,
    rng: random.Random,
    fr: dict,
    to: dict,
    layout: Optional[list] = None,
    target_shapes: Optional[list] = None,
) -> dict:
    first = board[fr["r"]][fr["c"]]
    second = board[to["r"]][to["c"]]
    powerups = []
    if is_powerup(first):
        powerups.append({"pos": fr, "partner": to, "cell": first})
    if is_powerup(second):
        powerups.append({"pos": to, "partner": fr, "cell": second})

    if not powerups:
        return {
            "total_score": 0,
            "chain_score": 0,
            "special_gained": {},
            "cleared_by_shape": {},
            "had_match": False,
            "used_powerup": False,
            "task_from_powerup": {},
        }

    total_score = 0
    combined_special: dict = {}
    combined_cleared: dict = {}
    triggered = False

    for pw in powerups:
        _targets, sp_gained, cleared = powerup_upgrade_targets(
            board, pw["pos"], pw["partner"], layout, target_shapes
        )
        if not _targets and pw["cell"].powerup_type not in ("row", "column", "color", "bomb"):
            continue
        triggered = True

        # bomb 逻辑：直接统计消除格数（_upgrade_targets 已经把格子删了）
        # row/column/color 逻辑：升级后 L2→L3 消失的格子也已经删了
        for shape, n in sp_gained.items():
            combined_special[shape] = combined_special.get(shape, 0) + n
        for shape, n in cleared.items():
            combined_cleared[shape] = combined_cleared.get(shape, 0) + n
        total_score += sum(cleared.values())

    if not triggered:
        return {
            "total_score": 0,
            "chain_score": 0,
            "special_gained": {},
            "cleared_by_shape": {},
            "had_match": False,
            "used_powerup": False,
            "task_from_powerup": {},
        }

    # 升级/消除后重力补位
    apply_gravity_and_refill(board, rng, layout)

    # 补位后可能触发新的普通三连
    chain = resolve_board(board, rng, fr, to, layout, target_shapes)
    total_score += chain["total_score"]

    for shape, n in chain["special_gained"].items():
        combined_special[shape] = combined_special.get(shape, 0) + n
    for shape, n in chain["cleared_by_shape"].items():
        combined_cleared[shape] = combined_cleared.get(shape, 0) + n

    return {
        "total_score": total_score,
        "chain_score": chain["total_score"],
        "special_gained": combined_special,
        "cleared_by_shape": combined_cleared,
        "merge_events": chain.get("merge_events", []),
        "had_match": chain["had_match"] or True,
        "used_powerup": True,
        # 任务分统一只走 special_gained，task_from_powerup 置空避免重复计分
        "task_from_powerup": {},
    }


def try_swap(
    board: Board,
    rng: random.Random,
    fr: dict,
    to: dict,
    layout: Optional[list] = None,
    target_shapes: Optional[list] = None,
) -> dict:
    swap_cells(board, fr, to)
    power_res = _resolve_powerup_swap(board, rng, fr, to, layout, target_shapes)
    if power_res["used_powerup"]:
        return power_res
    normal_res = resolve_board(board, rng, fr, to, layout, target_shapes)
    normal_res["task_from_powerup"] = {}
    return normal_res
